car-bridge/src/messaging.rs

```diff
@@ -0,0 +1,168 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+use std::time::Duration;
+
+use async_channel::Receiver;
+use async_stream::stream;
+use async_trait::async_trait;
+use chariott_common::{
+    config::env,
+    error::{Error, ResultExt as _},
+};
+use futures::{stream::BoxStream, StreamExt as _};
+use paho_mqtt::{
+    AsyncClient, ConnectOptionsBuilder, CreateOptionsBuilder, Message, MessageBuilder,
+    MQTT_VERSION_5, QOS_2,
+};
+use tracing::info;
+
+#[async_trait]
+pub trait Subscriber {
+    type Message;
+    type Topic;
+
+    /// Subscribe to a topic.
+    async fn subscribe(
+        &mut self,
+        topic: String,
+    ) -> Result<BoxStream<'static, Self::Message>, Error>;
```

Comment:
Why depend on Chariott's `Error`?

Reply:
We will need some error type. For now, I introduced an associated type to not require this as part of the trait definition with 2319d93fcb329ac9f71e8ae9c64ac74ba859d01e.

I kept reusing `chariott-common::error::Error` for the `MqttMessaging` implementation because I find it convenient. If I remove it, I would end up using `anyhow` (which is also just reusing another "common" error implementation) or creating an error type for the Car Bridge (which I don't think is worth it, as it would look identical to `chariott_common::error::Error`).

Reply:
It seems all your error are just string literals, so why not just use `&'static str` as your error type? Seems to be simple and works:

```diff
diff --git a/car-bridge/src/messaging.rs b/car-bridge/src/messaging.rs
index ecd3e513..5b54ebc6 100644
--- a/car-bridge/src/messaging.rs
+++ b/car-bridge/src/messaging.rs
@@ -6,10 +6,7 @@ use std::time::Duration;
 use async_channel::Receiver;
 use async_stream::stream;
 use async_trait::async_trait;
-use chariott_common::{
-    config::env,
-    error::{Error, ResultExt as _},
-};
+use chariott_common::config::env;
 use futures::{stream::BoxStream, StreamExt as _};
 use paho_mqtt::{
     AsyncClient, ConnectOptionsBuilder, CreateOptionsBuilder, Message, MessageBuilder,
@@ -58,7 +55,7 @@ impl MqttMessaging {
     /// If there was a persisted session before, messages may delivered before
     /// the `connect` returns. Refer to the `Subscriber` implementation for how
     /// to get access to the buffered messages.
-    pub async fn connect(client_id: String) -> Result<Self, Error> {
+    pub async fn connect(client_id: String) -> Result<Self, &'static str> {
         const BROKER_URL_ENV_NAME: &str = "BROKER_URL";
         const DEFAULT_BROKER_URL: &str = "tcp://localhost:1883";
         const MQTT_CLIENT_BUFFER_SIZE: usize = 200;
@@ -81,7 +78,7 @@ impl MqttMessaging {
                 .client_id(client_id)
                 .finalize(),
         )
-        .map_err_with("Failed to create MQTT client.")?;
+        .map_err(|_| "Failed to create MQTT client.")?;
 
         // Get the stream before connecting the client, as otherwise messages
         // may be lost. If the Car Bridge restarts while the number of
@@ -98,7 +95,7 @@ impl MqttMessaging {
                     .finalize(),
             )
             .await
-            .map_err_with("Could not connect to MQTT broker.")?;
+            .map_err(|_| "Could not connect to MQTT broker.")?;
 
         Ok(Self { client, receiver, is_subscribed: false })
     }
@@ -108,7 +105,7 @@ impl MqttMessaging {
 impl Subscriber for MqttMessaging {
     type Message = Message;
     type Topic = String;
-    type Error = Error;
+    type Error = &'static str;
 
     /// Subscribes to a topic on the MQTT broker. Currently, this can only be
     /// invoked once on `MqttMessaging`, due to the structure of the underlying
@@ -116,7 +113,7 @@ impl Subscriber for MqttMessaging {
     async fn subscribe<'a>(
         &'a mut self,
         topic: String,
-    ) -> Result<BoxStream<'static, Self::Message>, Error> {
+    ) -> Result<BoxStream<'static, Self::Message>, Self::Error> {
         // TODO: By broadcasting the events on the underlying `Receiver` and
         // filtering the broadcasted events by their topic name, we can support
         // multiple subscriptions. Since this is currently not needed, we do not
@@ -124,7 +121,7 @@ impl Subscriber for MqttMessaging {
         // instead.
 
         if self.is_subscribed {
-            return Err(Error::new("Already receiving messages. It is currently not possible to subscribe multiple times."));
+            return Err("Already receiving messages. It is currently not possible to subscribe multiple times.");
         }
 
         self.is_subscribed = true;
@@ -135,7 +132,7 @@ impl Subscriber for MqttMessaging {
         self.client
             .subscribe(topic.clone(), QOS_2)
             .await
-            .map_err_with("Could not subscribe to topic for receiving C2D messages.")?;
+            .map_err(|_| "Could not subscribe to topic for receiving C2D messages.")?;
 
         let mut receiver = self.receiver.clone();
 
@@ -160,13 +157,13 @@ impl Subscriber for MqttMessaging {
 impl Publisher for MqttMessaging {
     type Message = MessageBuilder;
     type Topic = String;
-    type Error = Error;
+    type Error = &'static str;
 
     /// Publish a message to an MQTT broker.
-    async fn publish(&self, topic: Self::Topic, message: Self::Message) -> Result<(), Error> {
+    async fn publish(&self, topic: Self::Topic, message: Self::Message) -> Result<(), Self::Error> {
         self.client
             .publish(message.topic(topic).finalize())
             .await
-            .map_err_with("Error when publishing a response.")
+            .map_err(|_| "Error when publishing a response.")
     }
 }
```

Reply:
It does not have the same semantics. When we use `map_err_with` we keep the inner error, but provide additional context through the error message. When we use a `&'static str`, we discard all additional context that we may get from an inner error.

Reply:
Ah, yes, I missed that in reading the code and had to refresh my memory of how `map_err_with` maintains the context.